### PR TITLE
Add the ability to go back in random selection history

### PR DESCRIPTION
**Pull request by EVAST9919**

### CODE REVIEW

File: osu.Game/Screens/Select/FooterButton.cs [hunk 1]
```diff
@@ -84,6 +84,7 @@ public FooterButton()
         public Action Hovered;
         public Action HoverLost;
         public Key? Hotkey;
+        public bool ExcludePressWithShift;
```
Comment (smoogipooo) [inline osu.Game/Screens/Select/FooterButton.cs]: I don't like this solution. Can the case where this is used (SelectRandom vs CancelRandom) not be handled inside SongSelect? Maybe rename the method and have something like...
```
var randomButton = new FooterButton(..., Keys.F2, TriggerRandom);

void TriggerRandom(InputState state)
{
    if (state.Keyboard.ShiftPressed)
        CancelRandom();
    else
        SelectRandom();
}
```

I don't like the naming of CancelRandom and SelectRandom either. Perhaps rename them to `SelectPreviousRandom` and `SelectNextRandom` respectively?

File: osu.Game/Screens/Select/BeatmapCarousel.cs [hunk 2]
```diff
@@ -197,6 +201,38 @@ public void SelectRandom()
...
+        {
+            if (!randomSelectedBeatmaps.Any())
+                return;
+
+            var visibleGroups = getVisibleGroups();
+            if (!visibleGroups.Any())
+                return;
+
+            // we can avoid selecting deleted beatmaps or beatmaps selected in another gamemode
+            while (true)
```
Comment (smoogipooo) [inline osu.Game/Screens/Select/BeatmapCarousel.cs]:
```
while (!randomSelectedBeatmaps.Count > 0)
{
    var group = randomSelectedBeatmaps.Pop();
    if (visibleGroups.Conains(group))
    {
        selectGroup(group, group.SelectPanel);
        break;
    }
}
```

**Review by smoogipoo — CHANGES_REQUESTED** [review]
As commented.

File: osu.Game/Screens/Select/BeatmapCarousel.cs [hunk 3]
```diff
@@ -77,8 +77,9 @@ public IEnumerable<BeatmapSetInfo> Beatmaps
 
         private readonly List<Panel> panels = new List<Panel>();
 
-        private BeatmapGroup selectedGroup;
+        private readonly Stack<BeatmapGroup> randomSelectedBeatmaps = new Stack<BeatmapGroup>();
```
Comment (smoogipooo) [inline osu.Game/Screens/Select/BeatmapCarousel.cs]: Am thinking - how about storing this as a `Stack<KeyValuePair<BeatmapGroup, BeatmapPanel>>` and having `SelectPreviousRandom` work for panels too? At the moment if I filter e.g. AiAe, and F2, I can't go back.

osu!stable does random by panel as I suggest also.

**Review by smoogipoo — APPROVED** [review]
I'm happy with the current state of this.

**Pull request merged**